html/user/per_app_list.php

```diff
@@ -56,74 +60,319 @@ function show_header($is_team, $apps, $appid, $is_total) {
+                        $fetch--;
+                      }
+             }
+             if ($check != $fetch) {
+                 $fetched_items++;
+                 $used_offset--;
+             }
+        } 
+        $store[0][0] = $fetch;
+        $store[1][0] = $fetched_items - $items_per_page;
         }
-        echo "<td align=right>".format_credit($c->expavg)."</td><td align=right>".format_credit_large($c->total)."</td>\n";
-    }
-    echo "</tr>\n";
+
+       return $data;
+
+}
+
+function choose_users($x, $appid, $used_offset, $up) {
+     global $store;
+     global $items_per_page;
+     $fetched_items = $items_per_page;
+     $check = 0;
+     $fetch = 1;
+  
+      //Do the same for the credit_user and user tables
+      //
+     if (!$up) {
+      while ($check != $fetch) {
+             $data = BoincCreditUser::enum("appid=$appid order by $x desc limit $used_offset,$fetched_items");
+             $check = sizeof($data) - ($fetched_items - $items_per_page);
+             $fetch = $check + ($fetched_items - $items_per_page);
+             foreach ($data as $item) {
+                      $user = BoincUser::lookup_id($item->userid);
+                      if (!is_object($user)) {
+                          $fetch--;
+                      }
+             }
+             if ($check != $fetch) {
+                   $fetched_items++;
+             }
+      }
+        $store[0][0] = $fetch;
+        $store[1][0] = $fetched_items - $items_per_page;
+     } else { 
+   
+       while ($check != $fetch) {
+              if ($used_offset < 0) $used_offset = 0;
+              $data =  BoincCreditUser::enum("appid=$appid order by $x desc limit $used_offset,$fetched_items");
+              $check = sizeof($data) - ($fetched_items - $items_per_page);
+              $fetch = $check + ($fetched_items - $items_per_page);
+              foreach ($data as $item) {
+                       $user = BoincUser::lookup_id($item->userid);
+                       if (!is_object($user)) {
+                         $fetch--;
+                       }
+              }
+              if ($check != $fetch) {
+                  $fetched_items++;
+                  $used_offset--;
+              }
+        }
+           $store[0][0] = $fetch;
+           $store[1][0] = $fetched_items - $items_per_page;
+       }
+        return $data;
 }
 
-function show_list($is_team, $appid, $is_total) {
-    $x = $is_team?"teams":"participants";
-    page_head("Top $x by application");
-    $apps = BoincApp::enum("deprecated=0");
-    if (!$appid) {
-        $appid = $apps[0]->id;
+function retrieve_credit_team($data) {
+     global $apps;
+     global $store;
+     $x = 1;
+     $c = 1;
+   
+     //Store the sign of each team in the first column of an array
+     // 
+     foreach ($data as $item) {
+                $team = BoincTeam::lookup_id($item->teamid);
+                if (is_object($team)) {
+                   $sign = "<td>".team_links($team)."</td>";
```

Comment:
Please don't store HTML in the cache. I know BOINC is not separating Logic and Presentation very good but HTML belongs into the `show_row()` function. The cache should only contain the raw values. This has the advantage that you don't need to invalidate the whole cache just because you tweaked the output a little bit.

In this occasion it seems ok to cache the output from `team_links()` (without the surrounding td) which is HTML just to avoid the Badge DB lookup when displaying a cached page. But for the other values I don't see the need to store anything but the raw values.